File: apps/android/app/src/main/java/fi/eport/searchassistant/data/session/SessionStore.kt
package fi.eport.searchassistant.data.session

import android.content.Context
import android.content.SharedPreferences

/// Per-slug auth tokens. SharedPreferences is fine here — these are
/// link-derived session secrets, not high-value credentials; the link-
/// trust threat model is identical to the iOS UserDefaults storage.
/// Keys mirror the iOS app and the web localStorage (`sa.session.{slug}`,
/// `sa.owner.{slug}`) so the prefix is easy to grep for.
class SessionStore(context: Context) {
    private val prefs: SharedPreferences =
        context.getSharedPreferences("session_store", Context.MODE_PRIVATE)

    // Session token (participant) -----------------------------------

    fun sessionToken(slug: String): String? = prefs.getString(sessionKey(slug), null)

    fun setSessionToken(token: String, slug: String) {
        prefs.edit().putString(sessionKey(slug), token).apply()
    }

    fun clearSessionToken(slug: String) {
        prefs.edit().remove(sessionKey(slug)).apply()
    }

    // Owner token (creator) ------------------------------------------

    fun ownerToken(slug: String): String? = prefs.getString(ownerKey(slug), null)

    fun setOwnerToken(token: String, slug: String) {
        prefs.edit().putString(ownerKey(slug), token).apply()
    }

    private fun sessionKey(slug: String) = "sa.session.$slug"
    private fun ownerKey(slug: String) = "sa.owner.$slug"
}

File: apps/ios/SearchAssistant/Services/SessionStore.swift
import Foundation

/// Per-slug auth tokens. UserDefaults is fine here — these are link-derived
/// session secrets, not high-value credentials; users are on the link-trust
/// model anyway. Keys mirror the web app's localStorage (`sa:session:{slug}`,
/// `sa:owner:{slug}`) so the prefix is easy to grep for if needed.
final class SessionStore {
    static let shared = SessionStore()
    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    // MARK: - Session token (participant)

    func sessionToken(for slug: String) -> String? {
        defaults.string(forKey: Self.sessionKey(slug))
    }

    func setSessionToken(_ token: String, for slug: String) {
        defaults.set(token, forKey: Self.sessionKey(slug))
    }

    func clearSessionToken(for slug: String) {
        defaults.removeObject(forKey: Self.sessionKey(slug))
    }

    // MARK: - Owner token (creator)

    func ownerToken(for slug: String) -> String? {
        defaults.string(forKey: Self.ownerKey(slug))
    }

    func setOwnerToken(_ token: String, for slug: String) {
        defaults.set(token, forKey: Self.ownerKey(slug))
    }

    // MARK: - Keys

    private static func sessionKey(_ slug: String) -> String { "sa.session.\(slug)" }
    private static func ownerKey(_ slug: String) -> String   { "sa.owner.\(slug)" }
}
